rustls-libssl/src/evp_pkey.rs

```diff
@@ -0,0 +1,231 @@
+            .and_then(|_| ctx.set_rsa_padding(RSA_PKCS1_PADDING))
+    }
+}
+
+unsafe impl Sync for RsaPkcs1 {}
+unsafe impl Send for RsaPkcs1 {}
+
+pub fn rsa_pss_sha256() -> Box<dyn EvpScheme + Send + Sync> {
+    Box::new(RsaPss(unsafe { EVP_sha256() }))
+}
+
+pub fn rsa_pss_sha384() -> Box<dyn EvpScheme + Send + Sync> {
+    Box::new(RsaPss(unsafe { EVP_sha384() }))
+}
+
+pub fn rsa_pss_sha512() -> Box<dyn EvpScheme + Send + Sync> {
+    Box::new(RsaPss(unsafe { EVP_sha512() }))
+}
+
+#[derive(Debug)]
+struct RsaPss(*const EVP_MD);
+
+impl EvpScheme for RsaPss {
+    fn digest(&self) -> *mut EVP_MD {
+        self.0 as *mut EVP_MD
+    }
+
+    fn configure_ctx(&self, ctx: &mut SignCtx) -> Option<()> {
+        const RSA_PSS_SALTLEN_DIGEST: c_int = -1;
+        ctx.set_signature_md(self.0)
+            .and_then(|_| ctx.set_rsa_padding(RSA_PKCS1_PSS_PADDING))
+            .and_then(|_| ctx.set_pss_saltlen(RSA_PSS_SALTLEN_DIGEST))
+    }
+}
+
+unsafe impl Sync for RsaPss {}
+unsafe impl Send for RsaPss {}
+
+/// Owning wrapper for a signing `EVP_MD_CTX`
+pub(crate) struct SignCtx {
+    md_ctx: *mut EVP_MD_CTX,
+    // owned by `md_ctx`
+    pkey_ctx: *mut EVP_PKEY_CTX,
+}
+
+impl SignCtx {
+    fn new(md: *mut EVP_MD, pkey: *mut EVP_PKEY) -> Option<Self> {
+        let md_ctx = unsafe { EVP_MD_CTX_new() };
+        let mut pkey_ctx = ptr::null_mut();
+
+        match unsafe { EVP_DigestSignInit(md_ctx, &mut pkey_ctx, md, ptr::null_mut(), pkey) } {
+            1 => {}
+            _ => {
+                unsafe { EVP_MD_CTX_free(md_ctx) };
+                return None;
+            }
+        };
+
+        Some(SignCtx { md_ctx, pkey_ctx })
+    }
+
+    fn set_signature_md(&mut self, md: *const EVP_MD) -> Option<()> {
+        unsafe { EVP_PKEY_CTX_set_signature_md(self.pkey_ctx, md) == 1 }.then_some(())
+    }
+
+    fn set_rsa_padding(&mut self, pad: c_int) -> Option<()> {
+        unsafe { EVP_PKEY_CTX_set_rsa_padding(self.pkey_ctx, pad) == 1 }.then_some(())
+    }
+
+    fn set_pss_saltlen(&mut self, saltlen: c_int) -> Option<()> {
+        unsafe { EVP_PKEY_CTX_set_rsa_pss_saltlen(self.pkey_ctx, saltlen) == 1 }.then_some(())
+    }
+
+    fn sign(self, data: &[u8]) -> Result<Vec<u8>, ()> {
+        // determine length
+        let mut len = 0;
+        match unsafe {
+            EVP_DigestSign(
+                self.md_ctx,
+                ptr::null_mut(),
+                &mut len,
+                data.as_ptr(),
+                data.len(),
+            )
+        } {
+            1 => {}
+            _ => return Err(()),
+        };
+
+        // do the signature
+        let mut out = vec![0u8; len];
+
+        match unsafe {
+            EVP_DigestSign(
+                self.md_ctx,
+                out.as_mut_ptr(),
+                &mut len,
+                data.as_ptr(),
+                data.len(),
+            )
+        } {
+            1 => {}
+            _ => return Err(()),
+        }
+
+        out.truncate(len);
```

Comment:
Is this a defensive measure? Maybe it should be a debug assert instead since we expect `len` to be consistent from the first `EVP_DigestSign` to populate `len` for building `out` on the stack, and the call to `EVP_DigestSign` to write the sig.

Reply:
Yeah, I was spooked by the man page:

> (...) Otherwise the **maximum        necessary size** of the output buffer is written to the siglen parameter. If sig is not NULL then before the call the siglen       parameter should contain the length of the sig buffer. If the call is successful the signature is written to sig and the       **amount of data written to siglen**.

That sounded to me like the maximum and actual amount could be different, and I think that is potentially true for some variable-size signature encodings.

Reply:
Good catch, that makes sense to me. Maybe worth a comment at the `truncate` site since it's a little bit subtle?

Reply:
I have renamed `len` to two variables: the first `max_len` for the first call, then `actual_len` for the second.

Reply:
WFM!